content_handler/runtime_holder.cc

```diff
@@ -129,43 +139,45 @@ blink::UnzipperProvider RuntimeHolder::GetUnzipperProviderForRootBundle() {
   };
 }
 
-void RuntimeHolder::DidCreateFramebuffer(
-    mojo::InterfaceHandle<mojo::Framebuffer> framebuffer,
-    mojo::FramebufferInfoPtr info) {
-  viewport_metrics_->physical_width = info->size->width;
-  viewport_metrics_->physical_height = info->size->height;
-  if (runtime_)
+void RuntimeHolder::OnInvalidation(mozart::ViewInvalidationPtr invalidation,
+                                   const OnInvalidationCallback& callback) {
+  FTL_DCHECK(invalidation);
+  pending_invalidation_ = false;
+
+  // Apply view property changes.
+  if (invalidation->properties) {
+    view_properties_ = std::move(invalidation->properties);
+    viewport_metrics_->physical_width =
+        view_properties_->view_layout->size->width;
+    viewport_metrics_->physical_height =
+        view_properties_->view_layout->size->height;
+    viewport_metrics_->device_pixel_ratio =
+        view_properties_->display_metrics->device_pixel_ratio;
     runtime_->SetViewportMetrics(viewport_metrics_);
+  }
 
-  blink::Threads::Gpu()->PostTask(ftl::MakeCopyable([
-    rasterizer = rasterizer_.get(), framebuffer = std::move(framebuffer),
-    info = std::move(info)
-  ]() mutable {
-    rasterizer->SetFramebuffer(std::move(framebuffer), std::move(info));
-  }));
+  // Remember the scene version for rendering.
+  scene_version_ = invalidation->scene_version;
+
+  // TODO(jeffbrown): Flow the frame time through the rendering pipeline.
+  if (outstanding_requests_ >= kPipelineDepth) {
+    did_defer_frame_request_ = true;
+  } else {
+    ++outstanding_requests_;
+    BeginFrame();
+  }
+
+  callback.Run();
```

Comment:
In the case where we defer the BeginFrame call, should we defer running the callback as well?